ionoscloud/data_source_cdn_distribution.go

```diff
@@ -0,0 +1,201 @@
+										Type:        schema.TypeBool,
+										Description: "Enable or disable WAF to protect the upstream host.",
+										Computed:    true,
+									},
+									"geo_restrictions": {
+										Type:     schema.TypeList,
+										Computed: true,
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												"block_list": {
+													Type: schema.TypeSet,
+													Elem: &schema.Schema{
+														Type: schema.TypeString,
+													},
+													Computed: true,
+												},
+												"allow_list": {
+													Type: schema.TypeSet,
+													Elem: &schema.Schema{
+														Type: schema.TypeString,
+													},
+													Computed: true,
+												},
+											},
+										},
+									},
+									"rate_limit_class": {
+										Type:        schema.TypeString,
+										Description: "Rate limit class that will be applied to limit the number of incoming requests per IP.",
+										Computed:    true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		Timeouts: &resourceDefaultTimeouts,
+	}
+}
+
+func dataSourceCdnDistributionRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(services.SdkBundle).CdnClient
+
+	idValue, idOk := d.GetOk("id")
+	domainValue, domainOk := d.GetOk("domain")
+
+	id := idValue.(string)
+	domain := domainValue.(string)
+
+	if idOk && domainOk {
+		diags := diag.FromErr(errors.New("id and domain cannot be both specified in the same time"))
+		return diags
+	}
+	if !idOk && !domainOk {
+		diags := diag.FromErr(errors.New("please provide the distribution id or domain"))
+		return diags
+	}
+
+	var distribution cdn.Distribution
+	var err error
+
+	if idOk {
+		/* search by ID */
+		distribution, _, err = client.SdkClient.DistributionsApi.DistributionsFindById(ctx, id).Execute()
+		if err != nil {
+			diags := diag.FromErr(fmt.Errorf("an error occurred while fetching the distribution with ID %s: %w", id, err))
+			return diags
+		}
+	} else {
+		var results []cdn.Distribution
+
+		distributions, _, err := client.SdkClient.DistributionsApi.DistributionsGet(ctx).Execute()
+		if err != nil {
+			diags := diag.FromErr(fmt.Errorf("an error occurred while fetching container distributions: %w", err))
+			return diags
+		}
+
+		results = *distributions.Items
+		if domainOk {
+			partialMatch := d.Get("partial_match").(bool)
+
+			log.Printf("[INFO] Using data source for container registry by domain with partial_match %t and domain: %s", partialMatch, domain)
+
+			if distributions.Items != nil && len(*distributions.Items) > 0 {
+				var distributionsByDomain []cdn.Distribution
+
+				for _, distributionItem := range *distributions.Items {
+					if isValidDomain(distributionItem.Properties, domain, partialMatch) {
+						distributionsByDomain = append(distributionsByDomain, distributionItem)
+					}
+				}
+				if len(distributionsByDomain) > 0 {
```

Comment:
here we can return early on len(distributionsByDomain) == 0 and we can avoid the nested else.

Reply:
I think changing this should be done when we can change all data sources, as they are all similar